File: src/services/statistics.ts
import { invoke } from '@tauri-apps/api/core'

import type {
  AddedCardsStats,
  CardBreakdown,
  ReviewHistoryStats,
  TimeRange,
  TodayProgress,
} from '../types/statistics'

export function getTodayProgress(): Promise<TodayProgress> {
  return invoke<TodayProgress>('stats_get_today_progress')
}

export function getCardBreakdown(): Promise<CardBreakdown> {
  return invoke<CardBreakdown>('stats_get_card_breakdown')
}

export function getReviewHistory(range: TimeRange): Promise<ReviewHistoryStats> {
  return invoke<ReviewHistoryStats>('stats_get_review_history', { range })
}

export function getAddedCards(range: TimeRange): Promise<AddedCardsStats> {
  return invoke<AddedCardsStats>('stats_get_added_cards', { range })
}

File: src-tauri/src/controller/statistics.rs
use crate::interface::anki::AnkiError;
use crate::interface::statistics::{
    AddedCardsStats, CardBreakdown, ReviewHistoryStats, TimeRange, TodayProgress,
};
use crate::AppState;

#[tauri::command]
pub fn stats_get_today_progress(
    state: tauri::State<'_, AppState>,
) -> Result<TodayProgress, AnkiError> {
    state.statistics.today_progress()
}

#[tauri::command]
pub fn stats_get_card_breakdown(
    state: tauri::State<'_, AppState>,
) -> Result<CardBreakdown, AnkiError> {
    state.statistics.card_breakdown()
}

#[tauri::command]
pub fn stats_get_review_history(
    state: tauri::State<'_, AppState>,
    range: TimeRange,
) -> Result<ReviewHistoryStats, AnkiError> {
    state.statistics.review_history(range)
}

#[tauri::command]
pub fn stats_get_added_cards(
    state: tauri::State<'_, AppState>,
    range: TimeRange,
) -> Result<AddedCardsStats, AnkiError> {
    state.statistics.added_cards(range)
}